lib/shared/services/request_cache_service.dart

```diff
@@ -0,0 +1,102 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:hive/hive.dart';
+import 'package:http/http.dart' as http;
+
+class RequestCacheService<T> {
+  final T Function(Map<String, dynamic>) fromJson;
+  final Map<String, dynamic> Function(T) toJson;
+  final Duration cacheDuration;
+  final String networkCacheKey;
+  final http.Client httpClient;
+
+  RequestCacheService({
+    required this.fromJson,
+    required this.toJson,
+    this.cacheDuration = const Duration(minutes: 1),
+    this.networkCacheKey = 'network_cache',
+    http.Client? httpClient,
+  }) : httpClient = httpClient ?? http.Client();
+
+  /// Fetches data from the cache and API.
+  Stream<T> fetchData(String url) async* {
+    final box = await Hive.openBox(networkCacheKey);
+    final cacheKey = _generateCacheKey(url);
+    bool cacheEmitted = false;
+
+    try {
+      // Check for cached data
+      final cachedEntry = await box.get(cacheKey);
+      if (cachedEntry != null && cachedEntry is Map) {
+        final cachedMap = Map<String, dynamic>.from(cachedEntry);
+        final cachedTimestamp =
+            DateTime.parse(cachedMap['timestamp'] as String);
+        final cachedDataJson =
+            Map<String, dynamic>.from(cachedMap['data'] ?? {});
+        final cachedData = fromJson(cachedDataJson);
+        // Emit cached data
+        yield cachedData;
+        cacheEmitted = true;
+
+        final now = DateTime.now();
+        // Decide whether to fetch new data based on cache validity
+        if (now.difference(cachedTimestamp) < cacheDuration) {
+          // Cache is still valid, but we'll fetch new data to check for updates
+          // return; // Uncomment this line to skip fetching new data
+        }
```

Comment:
this can just be a parameter in `fetchData()` that defaults to fetching new data